File: client/python/mesh/log/nop.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
import inspect
import io
import traceback
from datetime import datetime
from io import StringIO

from mesh.environ import Mode
from mesh.log.types import Logger, Level


def stack_tuple() -> (str, str):
    if Mode.Nolog:
        return "", ""
    frame = inspect.stack()[4]
    index = frame.filename.rindex("/")
    name = frame.filename[0 if index < 0 else index + 1:]
    return name, frame.lineno


class NopLogger(Logger):

    def __init__(self):
        self.__level__ = Level.INFO

    def name(self) -> str:
        return 'mesh'

    def info(self, fmt: str, *args: object):
        self.write(Level.INFO, fmt, *args)

    def warn(self, fmt: str, *args: object):
        self.write(Level.WARN, fmt, *args)

    def error(self, fmt: str, *args: object):
        self.write(Level.ERROR, fmt, *args)

    def debug(self, fmt: str, *args: object):
        self.write(Level.DEBUG, fmt, *args)

    def fatal(self, fmt: str, *args: object):
        self.write(Level.FATAL, fmt, *args)

    def stack(self, fmt: str, *args: object):
        self.write(Level.STACK, fmt, *args)

    def writer(self) -> io.BytesIO:
        return io.BytesIO()

    def level(self, level: Level):
        self.__level__ = level

    def write(self, level: Level, fmt: str, args):
        if not level.match(self.__level__):
            return

        # discard log
        if Mode.Nolog.enable():
            return

        if not args or len(args) < 1:
            self.do_write(level, fmt)
            return

        if len(args) == 1 and isinstance(args[0], BaseException):
            self.do_write(level, fmt, args[0])
            return

        frags = (fmt if fmt else "").split("{}")
        placeholders = []
        throwable = False
        for arg in args:
            if isinstance(arg, BaseException):
                throwable = True
                continue
            if isinstance(arg, str):
                placeholders.append(arg)
                continue
            placeholders.append(f"{arg}")

        msg = StringIO()
        for idx in range(len(frags)):
            msg.write(frags[idx])
            if len(placeholders) > idx:
                msg.write(placeholders[idx])
                continue
            if idx < len(frags) - 1:
                msg.write('{}')

        self.do_write(level, msg.getvalue(), throwable)

    def do_write(self, level: Level, msg: str, e: bool = False):
        print(f"{datetime.now().strftime('%Y-%m-%d %H:%M:%S')} {level.name} {msg}{self.stack_tuple()}")
        if e:
            print(traceback.format_exc())

    @staticmethod
    def stack_tuple() -> str:
        if not Mode.Metrics.enable():
            return ""
        frame = inspect.stack()[4]
        index = frame.filename.rindex("/")
        name = frame.filename[0 if index < 0 else index + 1:]
        return f" {name}:{frame.lineno}"

File: client/golang/log/nop.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package log

import (
	"context"
	"fmt"
	"github.com/be-io/mesh/client/golang/macro"
	"io"
	"time"
)

func init() {
	Provide(nop)
}

var nop = func() Logger {
	return &std{context: macro.Context(), level: INFO}
}()

type std struct {
	context context.Context
	level   Level
}

func (that *std) GetTraceId() string {
	return ""
}

func (that *std) GetSpanId() string {
	return ""
}

func (that *std) GetTimestamp() int64 {
	return 0
}

func (that *std) Name() string {
	return "std"
}

func (that *std) Info0(format string, args ...interface{}) {
	that.Print(that.context, INFO, format, args...)
}

func (that *std) Warn0(format string, args ...interface{}) {
	that.Print(that.context, WARN, format, args...)
}

func (that *std) Error0(format string, args ...interface{}) {
	that.Print(that.context, ERROR, format, args...)
}

func (that *std) Debug0(format string, args ...interface{}) {
	that.Print(that.context, DEBUG, format, args...)
}

func (that *std) Fatal0(format string, args ...interface{}) {
	that.Print(that.context, FATAL, format, args...)
}

func (that *std) Stack0(format string, args ...interface{}) {
	that.Print(that.context, STACK, format, args...)
}

func (that *std) Info(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, INFO, format, args...)
}

func (that *std) Warn(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, WARN, format, args...)
}

func (that *std) Error(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, ERROR, format, args...)
}

func (that *std) Debug(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, DEBUG, format, args...)
}

func (that *std) Fatal(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, FATAL, format, args...)
}

func (that *std) Stack(ctx context.Context, format string, args ...interface{}) {
	that.Print(ctx, STACK, format, args...)
}

func (that *std) Catch(err error) {
	if nil != err {
		that.Print(that.context, ALL, err.Error())
	}
}

func (that *std) Panic(err error) {
	if nil != err {
		that.Print(that.context, ALL, err.Error())
		panic(err)
	}
}

func (that *std) Writer() io.Writer {
	return that
}

func (that *std) Level(level Level) {
	that.level = level
}

func (that *std) Print(ctx context.Context, level Level, format string, args ...interface{}) {
	ltx, ok := ctx.(Context)
	if !ok {
		ltx = that
	}
	if !level.Is(int(that.level)) {
		return
	}
	params, _ := Split0(ctx, args)
	msg := fmt.Sprintf(format, params...)
	ref := Caller(4)
	fmt.Println(mof.Format(ltx, time.Now(), level, msg, ref))
}

func (that *std) Write(text []byte) (n int, err error) {
	fmt.Print(string(text))
	return len(text), nil
}
